feat(core): support contenteditable inputs

aktanoff opened the pull request:
Closes # https://github.com/taiga-family/maskito/issues/958

aktanoff commented on projects/core/src/lib/constants/default-element-predicate.ts:
```diff
@@ -1,5 +1,13 @@
-    (e as HTMLInputElement | HTMLTextAreaElement);
+export const MASKITO_DEFAULT_ELEMENT_PREDICATE: MaskitoElementPredicate = e => {
+    if (e.isContentEditable) {
+        return new MaskitoReactContentEditableElement(e).proxy;
```
maybe it shouldn't be used here, we can try to add this logic inside mask.ts if you want it to be working everywhere

aktanoff commented on projects/demo/src/pages/recipes/contenteditable/contenteditable-doc.component.ts:
```diff
@@ -0,0 +1,25 @@
+import {ChangeDetectionStrategy, Component} from '@angular/core';
```
i have zero experience in angular, so i just copied TextArea recipe page, and changed <textarea> to <div contenteditable="true">

aktanoff commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+        this.host = host;
+
+        // Proxy used to forward properties from HTMLElement to our instance
+        this.proxy = new Proxy(this, {
```
i didn't found any possible ways to automaticly forward all properties to host element except declared in MaskitoReactContentEditableElement, so i just added Proxy instance for this purpose

nsbarsukov commented:
@aktanoff Hello!

Sorry for late response!
I see you contribution but I will review it later (I need to firstly fix some issues in another project Taiga UI).
I will be back with feedback later.

aktanoff commented:
@nsbarsukov, hi, could you please take a look at this PR?

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+                if (prop in this) {
+                    const field = this[prop as keyof typeof this];
+
+                    return typeof field === 'function'
+                        ? (...args: unknown[]) => field.apply(this, args)
+                        : field;
```
Should we `Reflect` ? 🤔

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+
+                const hostProp = prop as keyof typeof this.host;
+
+                if (isWritable(this.host, hostProp)) {
```
Could `Reflect` helps to drop this complex logic ?

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+    return Boolean(descriptor && descriptor.writable);
+}
+
+export class MaskitoReactContentEditableElement {
```
Unfortunately, we can't extends `Proxy` 😢 
```ts
class MaskitoContentEditable extends Proxy {
    constructor(private readonly element: HTMLElement) {
        super(this.element, /* Proxy traps */)
    }
}
```

___

Hovewer, we do something like this:
```ts
class MaskitoContentEditable implements MaskitoElement {
  constructor(private readonly element: HTMLElement) {
    const proxyHost = this;

    return new Proxy(element, {/* Proxy traps */})
  }
}
```

I wonder if it is not hacky too much 🤔

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
Let's move this file to just `projects/core/src/lib/classes/content-editable.ts`

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+            return;
+        }
+
+        const range = document.createRange();
```
Using global `document` is not SSR-friendly.
Use 
```ts
element.ownerDocument
```

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    private get selection(): {from: number | null; to: number | null} {
+        const selection = getSelection();
```
```ts
this.element.ownerDocument.getSelection()
```

(SSR-friendly way)

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+        this.host.textContent = value;
+    }
+
+    private get selection(): {from: number | null; to: number | null} {
+        const selection = getSelection();
+
+        if (!selection) {
+            return {from: null, to: null};
+        }
+
+        const from = Math.min(selection.anchorOffset, selection.focusOffset);
+        const to = Math.max(selection.anchorOffset, selection.focusOffset);
+
+        return {from, to};
```
Create `projects/core/src/lib/utils/dom/get-content-editable-selection.ts`:

```ts
import {SelectionRange} from '../../types';

export function getContentEditableSelection(element: HTMLElement): SelectionRange {
    const selection = element.ownerDocument.getSelection();

    if (!selection) {
        return [0, 0];
    }

    const from = Math.min(selection.anchorOffset, selection.focusOffset);
    const to = Math.max(selection.anchorOffset, selection.focusOffset);

    return [from, to];
}

```

nsbarsukov commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+        return this.selection.to;
+    }
+
+    setSelectionRange(from: number | null, to: number | null): void {
+        if (from === null || to === null) {
+            return;
+        }
+
+        const range = document.createRange();
+
+        const maxRangeIndex = this.value.length;
+
+        range.setStart(this.host.firstChild || this.host, Math.min(from, maxRangeIndex));
+        range.setEnd(this.host.lastChild || this.host, Math.min(to, maxRangeIndex));
+        const selection = getSelection();
+
+        if (selection) {
+            selection.removeAllRanges();
+            selection.addRange(range);
+        }
+    }
```
Create `projects/core/src/lib/utils/dom/set-content-editable-selection.ts`:

```ts
import {SelectionRange} from '../../types';

export function setContentEditableSelection(
    element: HTMLElement,
    [from, to]: SelectionRange,
): void {
    const document = element.ownerDocument;
    const range = document.createRange();

    range.setStart(element.firstChild || element, from);
    range.setEnd(element.lastChild || element, to);
    const selection = document.getSelection();

    if (selection) {
        selection.removeAllRanges();
        selection.addRange(range);
    }
}

```

aktanoff commented on projects/core/src/lib/classes/adapters/content-editable-element.ts:
```diff
@@ -0,0 +1,110 @@
+    return Boolean(descriptor && descriptor.writable);
+}
+
+export class MaskitoReactContentEditableElement {
```
if we will have limited count of supported fields (MaskitoElement), we can have implementation of every field without using proxy

nsbarsukov replied:
@aktanoff idea to use `Proxy` is good enough for plugins

https://maskito.dev/core-concepts/plugins

aktanoff commented:
@nsbarsukov, Hello!

Sorry for the long absence, i've fixed all your issues in this iteration. And also added multiline masks support, the easiest way to implement it been using `white-space: pre;` style, and putting value as `string Node` into contenteditable element.

I've also seen your draft, and copied some code from there.

nsbarsukov commented on projects/core/src/lib/classes/content-editable.ts:
```diff
@@ -0,0 +1,71 @@
+import {setContentEditableSelection} from '../utils/dom/set-content-editable-selection';
+
+export class MaskitoContentEditable implements MaskitoElement {
+    // declared fields
+    declare addEventListener: HTMLElement['addEventListener'];
+    declare removeEventListener: HTMLElement['removeEventListener'];
+
+    declare dispatchEvent: HTMLElement['dispatchEvent'];
+    declare isContentEditable: HTMLElement['isContentEditable'];
+    declare nodeName: HTMLElement['nodeName'];
+    declare matches: HTMLElement['matches'];
+    declare ownerDocument: HTMLElement['ownerDocument'];
```
Could we improve solution somehow and don't iterate all possible properties ?
Actually, plugins can use ANY possible property of `HTMLElement`

nsbarsukov replied:
Unfortunately, without fixing this problem we can't merge this solution.
Our users could write their own plugins and use ANY possible properties of `HTMLElement`.
It will cause breaking change for them(

nsbarsukov commented on projects/core/src/lib/classes/content-editable.ts:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    get maxLength(): number {
+        return -1;
```
```suggestion
        return Infinity;
```

nsbarsukov commented on projects/demo/src/app/constants/demo-path.ts:
```diff
@@ -20,6 +20,7 @@ export const DemoPath = {
     Card: 'recipes/card',
     Phone: 'recipes/phone',
     Textarea: 'recipes/textarea',
+    ContentEditable: 'recipes/contenteditable',
```
```suggestion
    ContentEditable: 'recipes/content-editable',
```

nsbarsukov commented:
Closing this PR in favor of https://github.com/taiga-family/maskito/pull/1039
